[14.0][MIG] account_invoice_date_due: Migration to 14.0

joao-p-marques opened the pull request:
@Tecnativa 
TT28729

ping @pedrobaeza @victoralmau @chienandalu

luistorresm committed:
[ADD] account_invoice_date_due: Date Due now could be updated

Allow update the date due in an invoice (Even if it's done). To avoid
cancel the invoice and generate again

luisg123v committed:
[FIX] account_invoice_date_due: Small fixes to the initial version

This applies pending comments on #632, which are:
- Fix some typos
- Be consistent with the license, use AGPL-3
- Add maintainer (not commented)
- Set group data as noupdate (not commented)

marylla committed:
Added translation using Weblate (German)

marylla committed:
Translated using Weblate (German)

Currently translated at 100.0% (15 of 15 strings)

Translation: account-invoicing-12.0/account-invoicing-12.0-account_invoice_date_due
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-12-0/account-invoicing-12-0-account_invoice_date_due/de/

anasuarez1 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (15 of 15 strings)

Translation: account-invoicing-12.0/account-invoicing-12.0-account_invoice_date_due
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-12-0/account-invoicing-12-0-account_invoice_date_due/es/

joao-p-marques committed:
[IMP] account_invoice_date_due: black, isort, prettier

joao-p-marques committed:
[MIG] account_invoice_date_due: Migration to 13.0

TT28875

joao-p-marques committed:
[IMP] account_invoice_date_due: Refactor and add tests

Don't use the wizard, but make the invoice_date_due editable for a specific group
Still, propagate the change to the invoice lines
Add tests

TT28875

joao-p-marques committed:
[IMP] account_invoice_date_due: black, isort, prettier

chienandalu commented on account_invoice_date_due/models/account_move.py:
```diff
@@ -17,6 +17,22 @@ def _onchange_invoice_date_due_payment_term(self):
+    def onchange(self, values, field_name, field_onchange):
+        obj = self
+        if field_name == "invoice_date_due":
+            # A change in the invoice date due would trigger a recompute of the
```
Should we constraint this to the `posted` state?

joao-p-marques replied:
I was just writing a question about this :smile: 

I had some issues due to the fact that the change of the date due after the invoice had been posted triggered a recompute of the amounts in the lines and was marking the invoice as paid. I don't know if this is the correct fix though, as what I observed was that the residual amount was being computed as 0.0 after the change of the due date (See https://github.com/odoo/odoo/blob/14.0/addons/account/models/account_move.py#L1316)

The constraint you mention makes sense, though, if we follow this approach.

chienandalu replied:
I'd say so. Otherwise you're blocking the expected standard behavior. After that, as in core it's not possible to change the due date from the invoice, we must prevent to trigger that recompute: https://github.com/odoo/odoo/blob/14.0/addons/account/models/account_move.py#L491

joao-p-marques replied:
Ok, I made the suggested change of limiting to the `posted` state.

I also pushed another commit to fix the tests of `account_billing`, that were failing due to the change in Odoo from a couple of weeks ago of limiting the journal types for the specific invoice types.

joao-p-marques committed:
[MIG] account_invoice_date_due: Migration to 14.0

TT28729

pedrobaeza commented:
Can you please solve the conflict?

joao-p-marques commented:
@pedrobaeza done. I had a commit that fixed the tests in `account_billing`, but someone did it in the main branch as well, so I just removed it here.

pedrobaeza commented:
Travis is red

joao-p-marques committed:
[FIX] account_billing: Fix tests

After https://github.com/odoo/odoo/commit/fcaa54939e9a4f0dd5e47cd0ccffe7aa24bd451c, we need to set the invoice date manually.

joao-p-marques commented:
Thanks for the warning @pedrobaeza 
The tests in `account_billing` were failing after https://github.com/odoo/odoo/commit/fcaa54939e9a4f0dd5e47cd0ccffe7aa24bd451c
Should be fixed now.

pedrobaeza commented:
/ocabot merge nobump
